File: ws/bonifaz/u1/w08CDtoJava/ChickenFarmSimulator/src/utils/FileManager.java
package utils;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author Christian Bonifaz, Code Creators, DCCO-ESPE
 */
public class FileManager {

    public static void FileSave(String data, String fileName, int type) {

        if (type == 0) { //TXT
            fileName = fileName + ".txt";
        }
        if (type == 1) { //CSV
            fileName = fileName + ".csv";
        }
        if (type == 2) { //JSON
            fileName = fileName + ".json";

        }

        try (FileWriter fileWriter = new FileWriter(fileName, true); BufferedWriter bufferedWriter = new BufferedWriter(fileWriter)) {

            bufferedWriter.write(data);
            bufferedWriter.newLine();

        } catch (IOException e) {
            System.err.println("Error pepito writing to the file: " + e.getMessage());
        }
    }

    public static int FileReadLastId(String fileName, int type) {
        String lastLine = "";
        String separator = "";
        if (type == 0) { //TXT
            fileName = fileName + ".txt";
            separator = ":";
        }
        if (type == 1) { //CSV
            fileName = fileName + ".csv";
            separator = ",";
        }
        if (type == 2) { //JSON
            fileName = fileName + ".json";
            separator = ",";
        }
        try (BufferedReader read = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = read.readLine()) != null) {
                lastLine = line;
            }
            if (lastLine != null) {
                String[] values = lastLine.split(separator);
                try {
                    if (type == 2) {
                        return Integer.parseInt(values[0].split(":")[1]);
                    } else {
                        return Integer.parseInt(values[0]);
                    }

                } catch (NumberFormatException e) {
                    System.out.println("Error read last data");
                    return 0;
                }
            }

        } catch (IOException e) {
        }
        return 0;
    }

    public static void FileRead(String fileName, int type) {
        if (type == 0) { //TXT
            fileName = fileName + ".txt";
        }
        if (type == 1) { //CSV
            fileName = fileName + ".csv";
        }
        if (type == 2) { //JSON
            fileName = fileName + ".json";
        }
        try (BufferedReader read = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = read.readLine()) != null) {
                System.out.println(line);
            }

        } catch (IOException e) {
        }
    }

    public static String FindData(String fileName, int type, String wordSearch) {
        if (type == 0) {
            fileName = fileName + ".txt";
        }
        if (type == 1) {
            fileName = fileName + ".csv";
        }
        if (type == 2) {
            fileName = fileName + ".json";
        }
        try (BufferedReader bufferedReader = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = bufferedReader.readLine()) != null) {
                if (line.contains(wordSearch)) {
                    return line;
                }
            }
        } catch (IOException e) {
        }
        return "";
    }

    public static void UpdateData(String fileName, int type, String wordSearch, String newData) {
        
        if (type == 0) {
            fileName = fileName + ".txt";
        }
        if (type == 1) {
            fileName = fileName + ".csv";
        }
        if (type == 2) {
            fileName = fileName + ".json";
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.contains(wordSearch)) {
                    line = line.replace(wordSearch, newData);
                }
                FileSave(line, fileName, type);
            }
        } catch (IOException e) {
        }
    }



public static void DeleteData(String fileName, int type, int idEdit) {
        String separator = "";
        int numero;
        if (type == 0) { //TXT
            fileName = fileName + ".txt";
            separator = ":";
        }
        if (type == 1) { //CSV
            fileName = fileName + ".csv";
            separator = ",";
        }
        if (type == 2) { //JSON
            fileName = fileName + ".json";
            separator = ",";
        }
        List<String> lineas = new ArrayList<>();

        try {
            BufferedReader reader = new BufferedReader(new FileReader(fileName));
            String linea;
            while ((linea = reader.readLine()) != null) {
                String[] values = linea.split(separator);
                if (type == 2) {
                    numero = Integer.parseInt(values[0].split(":")[1]);
                } else {
                    numero = Integer.parseInt(values[0]);
                }
                if (numero == idEdit) {
                    continue;
                }
                lineas.add(linea);
            }
            reader.close();

            BufferedWriter writer = new BufferedWriter(new FileWriter(fileName));
            for (String l : lineas) {
                writer.write(l);
                writer.newLine();
            }
            writer.close();
        } catch (IOException e) {

        }
    }
}

File: hw/bonifaz/u1/hw09CDToJavaScrip/ChickenFarmSimulator/FileManager.js
import fs from 'fs';

export class FileManager {
    static FileSave(data, fileName, type) {
        let fileExtension = "";
        let separator = "";

        if (type === 0) { // TXT
            fileExtension = ".txt";
            separator = "\n";
        } else if (type === 1) { // CSV
            fileExtension = ".csv";
            separator = ",";
        } else if (type === 2) { // JSON
            fileExtension = ".json";
            separator = ",";
        }

        fileName += fileExtension;

        fs.appendFileSync(fileName, data + separator);
    }

    static FileReadLastId(fileName, type) {
        let lastLine = "";
        let separator = "";
        let lastId = 0;

        if (type === 0) { // TXT
            fileName += ".txt";
            separator = ":";
        } else if (type === 1 || type === 2) { // CSV or JSON
            fileName += (type === 1) ? ".csv" : ".json";
            separator = ",";
        }

        try {
            const data = fs.readFileSync(fileName, 'utf8');
            const lines = data.split('\n');

            if (lines.length > 0) {
                lastLine = lines[lines.length - 1];
                const values = lastLine.split(separator);

                if (type === 2) {
                    lastId = parseInt(values[0].split(":")[1]);
                } else {
                    lastId = parseInt(values[0]);
                }
            }
        } catch (error) {
            console.error("Error reading last ID:", error);
        }

        return lastId;
    }

    static FileRead(fileName, type) {
        if (type === 0) { // TXT
            fileName += ".txt";
        } else if (type === 1) { // CSV
            fileName += ".csv";
        } else if (type === 2) { // JSON
            fileName += ".json";
        }

        try {
            const data = fs.readFileSync(fileName, 'utf8');
            console.log(data);
        } catch (error) {
            console.error("Error reading file:", error);
        }
    }

    static FindData(fileName, type, wordSearch) {
        let separator = "";
        let result = "";

        if (type === 0) { // TXT
            fileName += ".txt";
        } else if (type === 1) { // CSV
            fileName += ".csv";
        } else if (type === 2) { // JSON
            fileName += ".json";
        }

        try {
            const data = fs.readFileSync(fileName, 'utf8');
            const lines = data.split('\n');

            for (const line of lines) {
                if (line.includes(wordSearch)) {
                    result = line;
                    break;
                }
            }
        } catch (error) {
            console.error("Error finding data:", error);
        }

        return result;
    }

    static UpdateData(fileName, type, wordSearch, newData) {
        let separator = "";

        if (type === 0) { // TXT
            fileName += ".txt";
        } else if (type === 1) { // CSV
            fileName += ".csv";
        } else if (type === 2) { // JSON
            fileName += ".json";
        }

        try {
            const data = fs.readFileSync(fileName, 'utf8');
            const lines = data.split('\n');

            const updatedLines = lines.map(line => {
                if (line.includes(wordSearch)) {
                    return line.replace(wordSearch, newData);
                }
                return line;
            });

            fs.writeFileSync(fileName, updatedLines.join('\n'));
        } catch (error) {
            console.error("Error updating data:", error);
        }
    }

    static DeleteData(fileName, type, idEdit) {
        let separator = "";
        let fileExtension = "";

        if (type === 0) { // TXT
            fileName += ".txt";
            separator = ":";
        } else if (type === 1) { // CSV
            fileName += ".csv";
            separator = ",";
        } else if (type === 2) { // JSON
            fileName += ".json";
            separator = ",";
        }

        let lines = [];

        try {
            const data = fs.readFileSync(fileName, 'utf8');
            lines = data.split('\n');
        } catch (error) {
            console.error("Error reading file:", error);
        }

        const filteredLines = lines.filter(line => {
            const values = line.split(separator);
            const numero = (type === 2) ? parseInt(values[0].split(":")[1]) : parseInt(values[0]);
            return numero !== idEdit;
        });

        try {
            fs.writeFileSync(fileName, filteredLines.join('\n'));
        } catch (error) {
            console.error("Error deleting data:", error);
        }
    }
}

export default FileManager; // Línea de exportación
